File: lua/tests/emu/emu_test.lua
local cpu = require("emu.cpu")
local tester = require("tests.tester")
local test_data = require("tests.emu.emu_test_data")

local emu_test = function(test_name, data)
	cpu.load_program(data.program)
	
	-- Mock io.write
	local output_buffer = ""
	local original_io_write = io.write
	io.write = function(str)
		output_buffer = output_buffer .. str
	end

	-- Safety cycle limit
	local cycles = data.max_cycles or 100
	cpu.run(cycles)

	-- Restore io.write
	io.write = original_io_write

	local success = true

	-- Check Output
	if data.expected_output and output_buffer ~= data.expected_output then
		print(string.format("Output mismatch in %s: Expected %q, got %q",
			test_name, data.expected_output, output_buffer))
		success = false
	end

	-- Check PC
	if data.expected_pc and cpu.pc ~= data.expected_pc then
		print(string.format("PC mismatch in %s: Expected %d, got %d",
			test_name, data.expected_pc, cpu.pc))
		success = false
	end

	-- Check Memory
	if data.expected_memory then
		-- addr in expected_memory is 0-based
		for addr, expected_val in pairs(data.expected_memory) do
			if cpu.memory[addr] ~= expected_val then
				print(string.format("Memory mismatch at address %d in %s: Expected %d, got %s",
					addr, test_name, expected_val, tostring(cpu.memory[addr])))
				success = false
			end
		end
	end

	return success
end

tester.run("Emulator Tests", test_data, emu_test)

File: js/tests/emu/emu_test.js
const cpu = require('../../emu/cpu');
const tester = require('../tester');
const testData = require('./emu_test_data.json');

const emuTest = (testName, data) => {
    cpu.load_program(data.program);

    let outputBuffer = "";
    const originalStdoutWrite = process.stdout.write;
    process.stdout.write = (str) => {
        outputBuffer += str;
        // Return true to indicate successful write (standard behavior)
        return true;
    };

    // Safety cycle limit
    const cycles = data.max_cycles || 100;
    cpu.run(cycles);

    // Restore stdout.write
    process.stdout.write = originalStdoutWrite;

    let success = true;

    // Check Output
    if (data.expected_output !== undefined && outputBuffer !== data.expected_output) {
        console.log(`Output mismatch in ${testName}: Expected ${JSON.stringify(data.expected_output)}, got ${JSON.stringify(outputBuffer)}`);
        success = false;
    }

    // Check PC
    if (data.expected_pc !== undefined && cpu.pc !== data.expected_pc) {
        console.log(`PC mismatch in ${testName}: Expected ${data.expected_pc}, got ${cpu.pc}`);
        success = false;
    }

    // Check Memory
    if (data.expected_memory) {
        for (const addr in data.expected_memory) {
            const expectedVal = data.expected_memory[addr];
            if (cpu.memory[addr] !== expectedVal) {
                console.log(`Memory mismatch at address ${addr} in ${testName}: Expected ${expectedVal}, got ${cpu.memory[addr]}`);
                success = false;
            }
        }
    }

    return success;
};

tester.run("Emulator Tests", testData, emuTest);
